File: tests/cocotb/zvfbf_test.cc
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <cstdint>

uint32_t vfwcvt_res[8] __attribute__((section(".data"))) = {0};
uint32_t vfncvt_res[8] __attribute__((section(".data"))) = {0};
uint32_t vfwmacc_res[8] __attribute__((section(".data"))) = {0};

int main() {
    // Test vfwcvtbf16.s.f.vv v8, v1
    // Source: v1 (BF16), Dest: v8 (FP32)
    // 1.5f (0x3fc0), 2.5f (0x4020), -1.0f (0xbf80), 0.0f (0x0000)
    uint32_t src_bf16[4] = {0x40203fc0, 0x0000bf80, 0, 0};

    asm volatile(
        "vsetvli t0, %[vl], e16, m1, ta, ma;"
        "vle16.v v1, (%[src]);"
        "vfwcvtbf16.f.f.v v8, v1;"
        "vse32.v v8, (%[dst]);"
        :
        : [src] "r"(src_bf16), [dst] "r"(vfwcvt_res), [vl] "r"(4)
        : "t0", "v1", "v8", "memory"
    );

    // Test vfncvtbf16.s.f.vv v1, v8
    // Source: v8 (FP32), Dest: v1 (BF16)
    asm volatile(
        "vsetvli t0, %[vl], e16, m1, ta, ma;"
        "vle32.v v8, (%[src]);"
        "vfncvtbf16.f.f.w v1, v8;"
        "vse16.v v1, (%[dst]);"
        :
        : [src] "r"(vfwcvt_res), [dst] "r"(vfncvt_res), [vl] "r"(4)
        : "t0", "v1", "v8", "memory"
    );

    // Test vfwmaccbf16.v.v.v v8, v1, v2
    // v8 = v8 + v1 * v2
    uint32_t src2_bf16[4] = {0x40004000, 0x40004000, 0, 0};
    asm volatile(
        "vsetvli t0, %[vl], e16, m1, ta, ma;"
        "vle16.v v1, (%[src1]);"
        "vle16.v v2, (%[src2]);"
        "vle32.v v8, (%[acc]);"
        "vfwmaccbf16.vv v8, v1, v2;"
        "vse32.v v8, (%[dst]);"
        :
        : [src1] "r"(src_bf16), [src2] "r"(src2_bf16), [acc] "r"(vfwcvt_res), [dst] "r"(vfwmacc_res), [vl] "r"(4)
        : "t0", "v1", "v2", "v8", "memory"
    );

    return 0;
}

File: tests/cocotb/zvfbf_test.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import cocotb
from coralnpu_test_utils.sim_test_fixture import Fixture
from bazel_tools.tools.python.runfiles import runfiles
import struct


def fp32_to_bits(f):
    return struct.unpack('<I', struct.pack('<f', f))[0]


def bits_to_fp32(i):
    return struct.unpack('<f', struct.pack('<I', i & 0xFFFFFFFF))[0]


@cocotb.test()
async def zvfbf_test(dut):
    """Test that runs Zvfbfmin and Zvfbfwma instructions."""
    fixture = await Fixture.Create(dut)
    r = runfiles.Create()
    elf_file = 'zvfbf_test.elf'

    await fixture.load_elf_and_lookup_symbols(
        r.Rlocation('coralnpu_hw/tests/cocotb/' + elf_file),
        ['vfwcvt_res', 'vfncvt_res', 'vfwmacc_res']
    )

    cycles = await fixture.run_to_halt(timeout_cycles=1000000)
    dut._log.info(f"Cycle count: {cycles}")

    # Verify vfwcvtbf16.s.f.vv
    # Expected: 1.5, 2.5, -1.0, 0.0
    expected_vfwcvt = [1.5, 2.5, -1.0, 0.0]
    dut._log.info("Verifying vfwcvt results...")
    for i in range(4):
        addr = fixture.symbols['vfwcvt_res'] + i * 4
        actual_bytes = await fixture.core_mini_axi.read(addr, 4)
        actual_bits = int.from_bytes(actual_bytes.tobytes(), 'little')
        actual_f = bits_to_fp32(actual_bits)
        dut._log.info(
            f"Element {i}: {actual_f} (expected {expected_vfwcvt[i]})"
        )
        assert actual_f == expected_vfwcvt[i]

    # Verify vfncvtbf16.s.f.vv
    # Expected: 0x3fc0, 0x4020, 0xbf80, 0x0000
    expected_vfncvt = [0x3fc0, 0x4020, 0xbf80, 0x0000]
    dut._log.info("Verifying vfncvt results...")
    for i in range(4):
        word_idx = i // 2
        elem_in_word = i % 2
        addr = fixture.symbols['vfncvt_res'] + word_idx * 4
        actual_bytes = await fixture.core_mini_axi.read(addr, 4)
        actual_word = int.from_bytes(actual_bytes.tobytes(), 'little')
        actual_bf16 = (actual_word >> (elem_in_word * 16)) & 0xFFFF
        dut._log.info(
            f"Element {i}: {hex(actual_bf16)} (expected {hex(expected_vfncvt[i])})"
        )
        assert actual_bf16 == expected_vfncvt[i]

    # Verify vfwmaccbf16.v.v.v
    expected_vfwmacc = [4.5, 7.5, -3.0, 0.0]
    dut._log.info("Verifying vfwmacc results...")
    for i in range(4):
        addr = fixture.symbols['vfwmacc_res'] + i * 4
        actual_bytes = await fixture.core_mini_axi.read(addr, 4)
        actual_bits = int.from_bytes(actual_bytes.tobytes(), 'little')
        actual_f = bits_to_fp32(actual_bits)
        dut._log.info(
            f"Element {i}: {actual_f} (expected {expected_vfwmacc[i]})"
        )
        assert actual_f == expected_vfwmacc[i]

    dut._log.info("All Vector BF16 tests passed!")
